File: VB/CustomTokenEdit/CustomTokenEditPainter.vb
Imports DevExpress.Skins
Imports DevExpress.Utils
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq

Namespace Example.CustomTokenEdit
	Public Class CustomTokenEditPainter
		Inherits TokenEditPainter

		Private editor As CustomTokenEdit
		Public Sub SetEditor(ByVal _editor As CustomTokenEdit)
			editor = _editor
		End Sub

		Protected Overrides Sub DoDrawGlyph(ByVal info As ControlGraphicsInfoArgs, ByVal tokenInfo As TokenEditTokenInfo)
			MyBase.DoDrawGlyph(info, tokenInfo)
			If editor.Properties.CheckBoxCollection.Images.Count <> 0 Then
				Dim index As Integer = 0
				Dim customTokenInfo As CustomTokenEditTokenInfo = TryCast(tokenInfo, CustomTokenEditTokenInfo)
				Dim offset As Integer = If(customTokenInfo.Checked = True, 4, 0)
				Select Case customTokenInfo.CheckBoxState
						Case ObjectState.Normal
							index = offset
						Case ObjectState.Hot
							index = offset + 1
						Case ObjectState.Hot Or ObjectState.Pressed
							index = offset + 2
				End Select
				Dim rect As Rectangle = customTokenInfo.CheckBoxGlyphBounds
				info.Graphics.DrawImage(editor.Properties.CheckBoxCollection.Images(index), New Rectangle(rect.X + CInt(Fix(info.Graphics.ClipBounds.X)), rect.Y + CInt(Fix(info.Graphics.ClipBounds.Y)), rect.Width, rect.Height))
			End If
		End Sub
	End Class
End Namespace

File: CS/TokenEdit_Glyph_InEndOfToken/CustomTokenEdit/CustomTokenEditPainter.cs
﻿using DevExpress.Skins;
using DevExpress.Utils;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;

namespace TokenEdit_Glyph_InEndOfToken.CustomTokenEdit {
    public class CustomTokenEditPainter : TokenEditPainter {
        CustomTokenEdit editor;
        public void SetEditor(CustomTokenEdit _editor) {
            editor = _editor;
        }

        protected override void DoDrawGlyph(ControlGraphicsInfoArgs info, TokenEditTokenInfo tokenInfo) {
            base.DoDrawGlyph(info, tokenInfo);

            if (editor.Properties.CheckBoxSkinElement != null) {
                int index = 0;
                CustomTokenEditTokenInfo customTokenInfo = tokenInfo as CustomTokenEditTokenInfo;
                int offset = customTokenInfo.Checked == true ? 4 : 0;
                Rectangle rect = customTokenInfo.CheckBoxGlyphBounds;
                Rectangle r = new Rectangle(rect.X + (int)info.Graphics.ClipBounds.X, rect.Y + (int)info.Graphics.ClipBounds.Y, rect.Width, rect.Height);
                SkinElementInfo skinElementInfo = new SkinElementInfo(editor.Properties.CheckBoxSkinElement, r);
                switch (customTokenInfo.CheckBoxState) {
                    case ObjectState.Normal:
                        index = offset;
                        break;
                    case ObjectState.Hot:
                        index = offset + 1;
                        break;
                    case ObjectState.Hot | ObjectState.Pressed:
                        index = offset + 2;
                        break;
                }
                skinElementInfo.ImageIndex = index;


                ImageCollection imageCollection = editor.Properties.CheckBoxSkinElement.Image.GetImages();
                info.Graphics.DrawImage(imageCollection.Images[index], new Rectangle(rect.X + (int)info.Graphics.ClipBounds.X, rect.Y + (int)info.Graphics.ClipBounds.Y, rect.Width, rect.Height));
                ObjectPainter.DrawObject(info.Cache, SkinElementPainter.Default, skinElementInfo);
            }
        }
    }
}
